src/NSSolver.cpp

```diff
@@ -0,0 +1,1033 @@
+                                       update_JxW_values);
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  FullMatrix<double> cell_pressure_mass_matrix(dofs_per_cell, dofs_per_cell);
+  Vector<double> cell_rhs(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  jacobian_matrix = 0.0;
+  residual_vector = 0.0;
+  pressure_mass = 0.0;
+
+  // Extractors
+  FEValuesExtractors::Vector velocity(0);
+  FEValuesExtractors::Scalar pressure(dim);
+
+  // We use these vectors to store the old solution (i.e. at previous Newton
+  // iteration) and its gradient on quadrature nodes of the current cell.
+  std::vector<Tensor<1, dim>> velocity_loc(n_q);
+  std::vector<Tensor<1, dim>> velocity_old_loc(n_q);
+  std::vector<Tensor<2, dim>> velocity_gradient_loc(n_q);
+  std::vector<double> pressure_loc(n_q);
+
+  // tensor used to store the nonlinear term contribution in each quadrature point
+  // useful when computing (u . nabla) uv, corresponding to c(u;u,v) term
+  Tensor<1, dim> nonlinear_term;
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+  {
+    if (!cell->is_locally_owned())
+      continue;
+
+    fe_values.reinit(cell);
+
+    cell_matrix = 0.0;
+    cell_rhs = 0.0;
+    cell_pressure_mass_matrix = 0.0;
+
+    // We need to compute the Jacobian matrix and the residual for current
+    // cell. This requires knowing the value and the gradient of u^{(k)}
+    // (stored inside solution) on the quadrature nodes of the current
+    // cell. This can be accomplished through
+    // FEValues::get_function_values and FEValues::get_function_gradients.
+    fe_values[velocity].get_function_values(solution, velocity_loc);
+    fe_values[velocity].get_function_gradients(solution,
+                                               velocity_gradient_loc);
+    fe_values[pressure].get_function_values(solution, pressure_loc);
+    fe_values[velocity].get_function_values(solution_old, velocity_old_loc);
+
+    for (unsigned int q = 0; q < n_q; ++q)
+    {
+      for (unsigned int i = 0; i < dofs_per_cell; ++i)
+      {
+        for (unsigned int j = 0; j < dofs_per_cell; ++j)
+        {
+          if (first_iter)
+          {
+            // Viscosity term.
+            cell_matrix(i, j) +=
+                nu *
+                scalar_product(fe_values[velocity].gradient(i, q),
+                               fe_values[velocity].gradient(j, q)) *
+                fe_values.JxW(q);
+
+            // Pressure term in the momentum equation.
+            cell_matrix(i, j) -= fe_values[velocity].divergence(i, q) *
+                                 fe_values[pressure].value(j, q) *
+                                 fe_values.JxW(q);
+
+            // time dependent term (mass matrix)
+            cell_matrix(i, j) += (velocity_loc[q] - velocity_old_loc[q]) *
+                                 fe_values[velocity].value(i, q) / delta_t *
+                                 fe_values.JxW(q);
```

Comment:
This looks incorrect to me: shouldn't you have the value of the j-th shape function in place of `velocity_loc[q]`, and shouldn't the term related to the old velocity go to th e right-hand side only?

Notice how this contribution to the (i,j)-th entry of the local matrix does not depend on j: something is definitely wrong here.